invokeai/app/authn.py

```
@@ -0,0 +1,97 @@
+import base64
+import binascii
+
+from fastapi import APIRouter
+from starlette import status
+from starlette.authentication import (
+    AuthCredentials,
+    AuthenticationBackend,
+    AuthenticationError,
+    SimpleUser,
+)
+from starlette.responses import HTMLResponse, PlainTextResponse
+
+from invokeai.app.services.config import InvokeAIAppConfig
+
+BASIC_AUTH_REALM = "InvokeAI"  # might want to make this configurable
+
+
+class AuthenticationRequired(AuthenticationError):
+    pass
+
+
+class BasicAuthBackend(AuthenticationBackend):
+    """Single-user username/password authentication."""
+
+    def __init__(self, app_config: InvokeAIAppConfig):
+        self._app_config = app_config
+        super().__init__()
+
+    async def authenticate(self, conn):
+        if self._app_config.password is None:
+            return None
+
+        # TODO: if connection is not secure, reject it as unsuitable for basic auth
+
+        if "Authorization" not in conn.headers:
+            raise AuthenticationRequired()
+
+        # Mostly copy/pasted from the starlette docs. It doesn't ship an implementation of this?
+        auth = conn.headers["Authorization"]
+        try:
+            scheme, credentials = auth.split()
+            if scheme.lower() != "basic":
+                # FIXME: this looks like its designed to fall-through to let another handler deal
+                #   with other schemes, but will that bypass our auth requirements? Or will the lack
+                #   of an authenticated user object prevent the request from getting through?
+                return
```

does this fail open or fail closed?